### PR TITLE
Initial checkin

### PR DESCRIPTION
**Pull request by ycombinator**
This PR introduces a first cut for this filter plugin.

The plugin converts string representations of computer storage sizes — for example, "123 MB" or "5.6Gb" — into their numeric value in bytes — for example, 128974848 and 700000000 respectively. It can also distinguish between bit and byte string units and perform the conversion appropriately.

This PR includes:
* [x] Code
* [x] Unit tests
* [x] Docs

### CODE REVIEW

**Comment by ycombinator** [comment]
Re: docs, some comments from @jakelandis and @dedemorton off-PR that I'm mentioning here mostly so I don't forget them:

* [x] Make sure you look at an existing plugin to see how the file is structured and how we do IDs.
* ~[ ] You may want to call out the ES bytes data type for the field mapping so that Kibana will re-format it for automatically~

**Comment by ycombinator** [comment]
@jakelandis said:

>  You may want to call out the ES bytes data type for the field mapping so that Kibana will re-format it for automatically

I'm not seeing this data type in ES docs. I see [`byte`](https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html) but that's something else. Could you point me in the right direction here please? Thanks.

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,2 @@
+source 'https://rubygems.org'
+gemspec
```
Comment (jakelandis) [inline Gemfile:2]: Can you add this snippet below this: https://github.com/logstash-plugins/logstash-input-beats/blob/master/Gemfile#L5-L11

This allows for testing against different branch versions.

Comment (jakelandis) [reply Gemfile:2]: Also, can you add a .travis file like this: https://github.com/logstash-plugins/logstash-input-beats/blob/master/.travis.yml

File: lib/logstash/filters/bytes.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+require "logstash/filters/base"
+require "logstash/namespace"
+
+# This filter will parse a given string as a computer storage
+# value (e.g. "123 MB" or "5.3GB") and add a new numeric field with
+# size in bytes
+#
+class LogStash::Filters::Bytes < LogStash::Filters::Base
+
+  @@prefix_powers = {
```
Comment (jakelandis) [inline lib/logstash/filters/bytes.rb]: should probably avoid global scope here in favor of a frozen const:
```
PREFIX_POWERS = {
...
}.freeze
```

File: lib/logstash/filters/bytes.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    if !match
+      event.set(@target, 0)
+      return
+    end
+
+    number, prefix, base = match.captures
+    # Take out any non-numeric characters or periods (e.g. delimiters) from
+    # the number
+    # TODO: i18n
+    number.tr!('^0-9.', '')
```
Comment (jakelandis) [inline lib/logstash/filters/bytes.rb]: Did you mean to leave the period in the expression ?

File: lib/logstash/filters/bytes.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    if prefix == '' && base == 'b'
+      result /= 8
+    end
+
+    # If prefix is uppercase (e.g. M) and base is 'b', then we've been given bits
+    if /[[:upper:]]/.match(prefix) && base == 'b'
+      result /= 8
+    end
+
+    # "kb" is a special case, we've been given bits
```
Comment (jakelandis) [inline lib/logstash/filters/bytes.rb]: is there a difference in `kb` vs `Kb` ? 

is there a difference in `mb` vs `Mb` ?

File: logstash-filter-bytes.gemspec [hunk 5]
```diff
@@ -0,0 +1,23 @@
+Gem::Specification.new do |s|
+  s.name          = 'logstash-filter-bytes'
+  s.version       = '0.1.0'
```
Comment (jakelandis) [inline logstash-filter-bytes.gemspec]: Suggest to go ahead and bump to 1.0.0

File: lib/logstash/filters/bytes.rb [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    source = event.get(@source)
+
+    if !source
+      return
+    end
+    source.strip!
+
+    # Parse the source into the number part (e.g. 123),
+    # the unit prefix part (e.g. M), and the unit base part (e.g. B)
+    match = source.match(/^([0-9\,\.]*)\s*([kKmMgGtTpPeE]?)([bB]?)$/)
```
Comment (jakelandis) [inline lib/logstash/filters/bytes.rb]: I imagine that rate as in `kb/s` is also a common pattern, may want to consider grabbing the post and placing it in a different field like <original_name>_PER as and translate `/s` to `second` ...of course that is a lot of scope creep.

File: DEVELOPER.md [hunk 7]
```diff
@@ -0,0 +1,2 @@
+# logstash-filter-bytes
+Example filter plugin. This should help bootstrap your effort to write your own filter plugin!
```
Comment (jakelandis) [inline DEVELOPER.md]: maybe remove this ?

File: spec/spec_helper.rb [hunk 8]
```diff
@@ -0,0 +1,2 @@
+# encoding: utf-8
```
Comment (jakelandis) [inline spec/spec_helper.rb]: you can probably just move this include to your spec for one less file to manage.

File: docs/index.asciidoc [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+[cols="<,<",options="header",]
+|=======================================================================
+|Input string |Numeric value in bytes
+|40|40
+|40B|40
+|40 B|40
+|40b|5
+|40KB|40960
+|40kB|40960
+|40.5KB|41472
```
Comment (jakelandis) [inline docs/index.asciidoc]: nit: this 40.5 example is bit out of place...the others are very nice.

File: docs/index.asciidoc [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+|40kb|5120
+|40Kb|5120
+|10 MB |10485760
+|10 mB |10485760
+|10 mb |10485760
+|10 Mb |1310720
+|=======================================================================
+
+[id="plugins-{type}s-{plugin}-options"]
+==== Bytes Filter Configuration Options
```
Comment (jakelandis) [inline docs/index.asciidoc]: A copy and paste example is usually SUPER nice to have.

**Comment by jakelandis** [comment]
@ycombinator  
>I'm not seeing this data type in ES docs. I see byte but that's something else. Could you point me in the right direction here please? Thanks.

Doh! I was mixing up ES mapping types with Kibana's number based formatting options .
image

File: README.md [hunk 11]
```diff
@@ -0,0 +1,86 @@
+# Logstash Plugin
+
+This is a plugin for [Logstash](https://github.com/elastic/logstash).
+
+It is fully free and fully open source. The license is Apache 2.0, meaning you are pretty much free to use it however you want in whatever way.
+
+## Documentation
+
+Logstash provides infrastructure to automatically generate documentation for this plugin. We use the asciidoc format to write documentation so any comments in the source code will be first converted into asciidoc and then into html. All plugin documentation are placed under one [central location](http://www.elastic.co/guide/en/logstash/current/).
```
Comment (dedemorton) [inline README.md]: This boiler plate text is no longer true. Comments in the source code are not converted into asciidoc. Is this boiler plate text defined somewhere so we can change it for developers who are contributing new LS plugins?

File: README.md [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+```
+- Run Logstash with your plugin
+```sh
+bin/logstash -e 'filter {awesome {}}'
+```
+At this point any modifications to the plugin code will be applied to this local Logstash setup. After modifying the plugin, simply rerun Logstash.
+
+#### 2.2 Run in an installed Logstash
+
+You can use the same **2.1** method to run your plugin in an installed Logstash by editing its `Gemfile` and pointing the `:path` to your local plugin development directory or you can build the gem and install it using:
```
Comment (dedemorton) [inline README.md:66]: "2.1 method" almost sounds like you are referring to a product version rather than another section in the document.  I'd change the wording around and maybe break this into 2 sentences:

You can run your plugin in an installed Logstash instance by editing the `Gemfile` and pointing the `:path` to your local plugin development directory (see the previous section). Or you can build the gem and install it by using: ...

File: docs/index.asciidoc [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+[id="plugins-{type}s-{plugin}"]
+
+=== Bytes filter plugin
+
+include::{include_path}/plugin_header.asciidoc[]
+
+==== Description
+
+Parse string representations of computer storage sizes such as "123 MB" or "5.6Gb" into
```
Comment (dedemorton) [inline docs/index.asciidoc]: Add commas to make this easier to parse: "....of computer storage sizes, such as "123 MB" or "5.6Gb", into..."

File: docs/index.asciidoc [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+&nbsp;
+
+[id="plugins-{type}s-{plugin}-source"]
+===== `source` 
+
+  * Value type is <<string,string>>
+  * Default value is `description`
+
+Name of source field from which to parse the storage size string.
```
Comment (dedemorton) [inline docs/index.asciidoc]: Maybe shorten this to improve readability: "Name of the source field that contains the storage size"

File: docs/index.asciidoc [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+[id="plugins-{type}s-{plugin}-target"]
+===== `target` 
+
+  * Value type is <<string,string>>
+
+Name of target field into which numeric value in bytes should be written.
```
Comment (dedemorton) [inline docs/index.asciidoc]: Some slight changes:

"Name of the target field that will contain the storage size in bytes"

**Review by dedemorton — COMMENTED** [review]
A few minor changes. Otherwise LGTM.

File: lib/logstash/filters/bytes.rb (hunk 3 above)
Comment (ycombinator) [reply lib/logstash/filters/bytes.rb]: Yes, for the decimal point. The comment above is unclear so I'll update that.

File: lib/logstash/filters/bytes.rb (hunk 4 above)
Comment (ycombinator) [reply lib/logstash/filters/bytes.rb]: Yes, I'm treating `mb`, `MB`, and `mB` as megabytes and `Mb` as megabits

I wanted to give the same treatment for kilobytes vs. kilobits but [apparently](https://en.wikipedia.org/wiki/Kilobit) `kb` is supposed to be kilobits. Hence the special casing here.

File: lib/logstash/filters/bytes.rb (hunk 6 above)
Comment (ycombinator) [reply lib/logstash/filters/bytes.rb]: Yeah, a future PR perhaps?

File: docs/index.asciidoc (hunk 9 above)
Comment (ycombinator) [reply docs/index.asciidoc]: Ah, I wanted to have an example that shows decimal usage.

File: README.md (hunk 11 above)
Comment (ycombinator) [reply README.md]: Yeah, it got generated when I ran the plugin generator: https://www.elastic.co/guide/en/logstash/current/_how_to_write_a_logstash_filter_plugin.html#_use_the_plugin_generator_tool_3.

I see this text in other plugins' READMEs. Should I keep it here for consistency with other plugins or delete it so it reflects the current state of the world accurately? I think the latter but want to confirm first.

File: README.md (hunk 12 above)
Comment (ycombinator) [reply README.md:66]: This is also part of the text generated by the plugin generator.

**Comment by ycombinator** [comment]
@jakelandis @dedemorton Thanks for the review. I've addressed all the feedback, either via changes or as replies to your comments. Ready for your 👀 again. Thanks!

File: README.md (hunk 11 above)
Comment (dedemorton) [reply README.md]: I would change the text. You can look at the PRs I have open here: https://github.com/logstash-plugins/logstash-input-example/pull/12

File: README.md (hunk 12 above)
Comment (dedemorton) [reply README.md:66]: Hmm...then I'd leave it as-is for now.

**Review by dedemorton — COMMENTED** [review]
Minor change to boiler plate text. Otherwise, LGTM.

File: README.md (hunk 11 above)
Comment (ycombinator) [reply README.md]: @dedemorton @jakelandis Since I generated this `README.md` using the plugin generator (`bin/logstash-plugin generate ...`), I think I should also make the suggested changes in:

* https://github.com/elastic/logstash/blob/master/lib/pluginmanager/templates/input-plugin/README.md#documentation
* https://github.com/elastic/logstash/blob/master/lib/pluginmanager/templates/filter-plugin/README.md#documentation,
* https://github.com/elastic/logstash/blob/master/lib/pluginmanager/templates/output-plugin/README.md#documentation, and
* https://github.com/elastic/logstash/blob/master/lib/pluginmanager/templates/codec-plugin/README.md#documentation

Otherwise any future plugin authors will end up with the wrong text. Correct?

File: Gemfile (hunk 1 above)
Comment (ycombinator) [reply Gemfile:2]: @jakelandis Should both these changes be made under https://github.com/elastic/logstash/tree/master/lib/pluginmanager/templates as well for future runs of the plugin generator (`bin/logstash-plugin generate ...`)?

**Comment by ycombinator** [comment]
@jakelandis and @dedemorton Some of the changes you suggested above should probably go into the plugin generator codebase itself, so future plugin authors automatically get them when the run `bin/logstash-plugin generate ...`. For that I've put up https://github.com/elastic/logstash/pull/9193 and requested your review on it. Thanks!

**Comment by ycombinator** [comment]
@dedemorton @jakelandis Also, this PR is ready for review again. Thanks!

**Review by dedemorton — APPROVED** [review]

File: lib/logstash/filters/bytes.rb [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+require "logstash/namespace"
+
+# This filter will parse a given string as a computer storage
+# value (e.g. "123 MB" or "5.3GB") and add a new numeric field with
+# size in bytes
+#
+class LogStash::Filters::Bytes < LogStash::Filters::Base
+
+  PREFIX_POWERS = {
+    'k' => 1, # 1 kilobyte = 1024 ^ 1 bytes
```
Comment (andrewvc) [inline lib/logstash/filters/bytes.rb:12]: The prefixes mean different things depending on whether we're talking network data or size on disk.  See: https://en.wikipedia.org/wiki/Mebibyte .

It'd be nice to also support `MiB` and `KiB`, even though their use is a little rare.

More important would be to add an option that toggles between SI Units and powers of 2.

Comment (andrewvc) [reply lib/logstash/filters/bytes.rb:12]: By SI units, I mean `k = 1000`

Comment (ycombinator) [reply lib/logstash/filters/bytes.rb:12]: > It'd be nice to also support MiB and KiB, even though their use is a little rare.

I'd like to do this in a follow up PR, if that's okay?

> More important would be to add an option that toggles between SI Units and powers of 2.

Agreed. I will add this option to this PR.

File: Gemfile (hunk 1 above)
Comment (jakelandis) [reply Gemfile:2]: Yes, and thank you for https://github.com/elastic/logstash/pull/9193

File: spec/filters/bytes_spec.rb [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+    end
+
+    sample("32eb") do
+      expect(subject).to include("dest")
+      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024 * 1024 * 1024)
+    end
+
+    sample("32EB") do
+      expect(subject).to include("dest")
+      expect(subject.get('dest')).to eq(32 * 1024 * 1024 * 1024 * 1024 * 1024 * 1024)
```
Comment (jakelandis) [inline spec/filters/bytes_spec.rb:151]: This is a really edge case (today)...but found this a while back with AWS EFS ... it is possible for exabytes to overflow a long when represented as bytes. Maybe we shouldn't support exabytes ?

Comment (ycombinator) [reply spec/filters/bytes_spec.rb:151]: Sure, that sounds safer. We can not support them until someone asks for them. Will remove. Thanks.

File: docs/index.asciidoc [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+|10 MB |10485760
+|10 mB |10485760
+|10 mb |10485760
+|10 Mb |1310720
+|=======================================================================
+
+[source,ruby]
+    filter {
+      bytes {
+        source => "my_bytes_string_field",
```
Comment (jakelandis) [inline docs/index.asciidoc]: shouldn't have a comma here (it wont compile with it)

**Comment by jakelandis** [comment]
I think the math is off when using a decimal point.

With the following config:
```
input { stdin {  }}    

filter {
  bytes {
        source => "message"
        target => "result"
      }
}
output {
  stdout { codec => rubydebug }
}
```
```
1.1KB
{
        "result" => 1126.4,
       "message" => "1.1KB",
      "@version" => "1",
          "host" => "macd",
    "@timestamp" => 2018-02-23T19:30:11.758Z
}
1.1MB
{
        "result" => 1153433.6,
       "message" => "1.1MB",
      "@version" => "1",
          "host" => "macd",
    "@timestamp" => 2018-02-23T19:30:19.701Z
}
```
EDIT:
Something is also funky with `1,1KB` vs. `1.1KB`
```
{
        "result" => 11264.0,
       "message" => "1,1KB",
      "@version" => "1",
          "host" => "macd",
    "@timestamp" => 2018-02-23T19:39:12.975Z
}
```

**Comment by jakelandis** [comment]
```
{
        "result" => 0,
       "message" => "junk",
      "@version" => "1",
          "host" => "macd",
    "@timestamp" => 2018-02-23T19:36:16.797Z
}
```
Rather then a zero if not parse-able, I think adding a tag is more Logstash idiomatic 

For example the json-filter if it can not parse the JSON will tag the event : https://github.com/logstash-plugins/logstash-filter-json/blob/master/lib/logstash/filters/json.rb#L60 This allows for the differentiation of a zero value vs. a failure.

**Comment by jakelandis** [comment]
I know in order to support bits, we need fractional bytes. , e.g. 1b = .125B , however for all other conversions it is a whole number.  

I am not sure the support for bit to byte conversion is worth the additional overhead of floating point type numbers in ES. https://www.elastic.co/guide/en/elasticsearch/reference/current/number.html#_which_type_should_i_use  

I doubt that many things are measured in bits, and loosing support for them may simplify things and be more efficient to use long types.  

Thoughts ?

EDIT:
Or keep support and just handle them specially and round up / down to nearest whole number.

**Comment by ycombinator** [comment]
@jakelandis Happy to drop bits for now and add in support later. As you pointed out they're probably not used much so we can wait till we get a request for them.

**Comment by ycombinator** [comment]
@jakelandis said:

> Something is also funky with 1,1KB vs. 1.1KB
>
>```
>{
>  "result" => 11264.0,
>  "message" => "1,1KB",
>  "@version" => "1",
>  "host" => "macd",
>  "@timestamp" => 2018-02-23T19:39:12.975Z
>}
>```

I think this is expected math. The plugin doesn't currently support `,` as an alternative to `.` (the decimal point). So `1,1KB` is treated as `11KB` which converts to 11 * 1024 = 11264 bytes.

[EDIT] What we could do, preferably in a follow up PR, is to support options such as `decimal_delimiter` and `digit_group_delimiter`. Thoughts?

**Comment by jakelandis** [comment]
@ycombinator - It looks like `.` is the minority , and `,` is the majority for decimal separators. https://en.wikipedia.org/wiki/Decimal_separator.  

I _think_ `1,1` is incorrect to assume the `,` is a group deliminator since it does not represent a hundred(th) grouping. `1,123` could be group or decimal, and this is where an explicit option would help to disambiguate . 

Is it possible to put in some small heuristics... if more then 1 `,` or `.` , it is a group deliminator, if it only 1 `,` or `.` AND the digits are not = 3 then it is  decimal deliminator...but if digits = 3 then use configurable option `decimal_delimiter` to disambiguate. 

I would consider ^^ to be close to a "must have" given our global community.  I am fine if that is a future PR, but should probably be addressed before this is a default plugin. 

Also, we should probably handle `_` and `<space>` group delimiters... i think we can do that auto-matically without exposing an explicit option assuming we expose `decimal_delimiter` to disambiguate the commas and periods for 3 based digits.

**Comment by ycombinator** [comment]
> I think 1,1 is incorrect to assume the , is a group deliminator since it does not represent a hundred(th) grouping

Yes, of course. Duh! :)

> Is it possible to put in some small heuristics... if more then 1 , or . , it is a group deliminator, if it only 1 , or . AND the digits are not = 3 then it is decimal deliminator...but if digits = 3 then use configurable option decimal_delimiter to disambiguate.

I like this approach! Will implement.

**Comment by ycombinator** [comment]
@jakelandis I thought about the heuristics-based approach some more and I think it might be trappy for our users (e.g. when I specify `1,333mb` do I mean `1333mb` or `1333kb`?). I'd rather give users _explicit_ options to override the [digits group separator](https://en.wikipedia.org/wiki/Decimal_separator#Digit_grouping) (defaulting to ` ` (a space)) and the [decimal separator](https://en.wikipedia.org/wiki/Decimal_separator) (defaulting to `.`). The defaults were chosen based on [current standards](https://en.wikipedia.org/wiki/Decimal_separator#Current_standards). WDYT?

**Comment by ycombinator** [comment]
Never mind @jakelandis, I just re-read your heuristic more carefully and I don't think its trappy. I've implemented it in 3f4802a.

This PR is ready for your 👀 again. Thanks!

File: docs/index.asciidoc [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+- petabytes ("PB", "pb", or "pB")
+
+[id="plugins-{type}s-{plugin}-examples"]
+==== Examples
+
+
+[cols="<,<",options="header",]
+|=======================================================================
+|Input string |Prefix system | Numeric value in bytes
+|40|`binary` or `decimal`|40
```
Comment (jakelandis) [inline docs/index.asciidoc]: the doc below states that this can be `binary` or `metric` ... but the doc here says `binary` or `decimal`

Comment (ycombinator) [reply docs/index.asciidoc]: Yeah, this should be `metric`. Good catch! Fixing.

File: docs/index.asciidoc [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+
+[id="plugins-{type}s-{plugin}-target"]
+===== `target` 
+
+  * Value type is <<string,string>>
+
+Name of the target field that will contain the storage size in bytes
+
+[id="plugins-{type}s-{plugin}-prefix_system"]
+===== `prefix_system`
```
Comment (jakelandis) [inline docs/index.asciidoc]: is `prefix_system` a standard term ?  if not can i suggest `conversion_method` or the like ? _(I understand functionality here.. but still don't understand what a prefix_system is)_

Comment (ycombinator) [reply docs/index.asciidoc]: No, I don't think its standard. I made it up after reading https://en.wikipedia.org/wiki/Metric_prefix and https://en.wikipedia.org/wiki/Binary_prefix. If it doesn't make sense I'm happy to change it; `conversion_method` seems better.

File: docs/index.asciidoc [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+[id="plugins-{type}s-{plugin}-prefix_system"]
+===== `prefix_system` 
+
+  * Value type is <<string,string>>
+  * Value can be any of: `binary`, `metric`
+  * Default value is `binary`
+
+Which prefix system to use when converting to bytes. `binary` uses `1K = 1024B`. `metric` uses `1K = 1000B`.
+
+[id="plugins-{type}s-{plugin}-decimal_separator"]
```
Comment (jakelandis) [inline docs/index.asciidoc]: nitpick: should it be decimal or fractional separator ?  is using a comma as the "decimal" separator a very American way to think about it ? (_presumes to use comma instead of the decimal_)

Comment (ycombinator) [reply docs/index.asciidoc]: I chose decimal separator because that's what Wikipedia seems to call it: https://en.wikipedia.org/wiki/Decimal_separator. I think we should leave this one as-is.

File: lib/logstash/filters/bytes.rb [hunk 22]
```diff
@@ -0,0 +1,151 @@
...
+  config_name "bytes"
+  
+  # Source string, e.g. "123 MB", "5.3GB", etc.
+  config :source, :validate => :string, :default => "message"
+
+  # Target field name
+  config :target, :validate => :string
+
+  # Conversion method, either "binary" (1K = 1024B) or "metric" (1K = 1000B)
+  config :conversion_method, :validate => :string, :default => "binary"
```
Comment (jakelandis) [inline lib/logstash/filters/bytes.rb]: you can add some validation to this like here: https://github.com/logstash-plugins/logstash-input-tcp/blob/master/lib/logstash/inputs/tcp.rb#L80

Comment (ycombinator) [reply lib/logstash/filters/bytes.rb]: Sweet!

File: lib/logstash/filters/bytes.rb [hunk 23]
```diff
@@ -0,0 +1,151 @@
...
+
+  # Decimal separator
+  config :decimal_separator, :validate => :string, :default => "."
```
Comment (jakelandis) [inline lib/logstash/filters/bytes.rb]: validation here too...

File: lib/logstash/filters/bytes.rb [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+  def register
+    # Add instance variables 
+  end # def register
+
+  public
+  def filter(event)
+
+    source = event.get(@source)
+
+    if !VALID_CONVERSION_METHODS.include?(@conversion_method)
```
Comment (jakelandis) [inline lib/logstash/filters/bytes.rb]: I think you can remove this since the config is doing the validation now.

**Comment by jakelandis** [comment]
It looks you dropped support for bits, but still output a float type:

`1.001kb -> 1025.024`

The fractional bits highlight some in-acuracy in the math w/r/t the fractional portions why doing base2 conversions.   For example, 1.001kb says 1024 bytes + 1 byte = 1025 , not 1025.024.  I think the accuracy here is acceptable, and getting the fractional accuracy for base2 is non-trival (i think). 

To avoid the mis-perception of high accuracy for fractional values, and no need for bits, can we either do whole number math, or fractional math then convert to long prior to the result ?

**Comment by ycombinator** [comment]
> To avoid the mis-perception of high accuracy for fractional values, and no need for bits, can we either do whole number math, or fractional math then convert to long prior to the result ?

Yeah, agreed. Let me see what works but the end result will be a long.

**Comment by ycombinator** [comment]
@jakelandis I decided to round the _result_ to the nearest whole number (as opposed to converting the _input_ to an integer). I think this will give better precision overall. WDYT? b8bebda

File: lib/logstash/filters/bytes.rb [hunk 25]
```diff
@@ -0,0 +1,145 @@
...
+
+    if !source
+      @tag_on_failure.each{|tag| event.tag(tag)}
+      return
+    end
+    source.strip!
+
+    # Parse the source into the number part (e.g. 123),
+    # the unit prefix part (e.g. M), and the unit suffix part (e.g. B)
+    match = source.match(/^([0-9#{DIGIT_GROUP_SEPARATORS}#{@decimal_separator}]*)\s*([kKmMgGtTpPeE]?)([bB]?)$/)
```
Comment (jakelandis) [inline lib/logstash/filters/bytes.rb:75]: Looks like match is still supporting Exabytes which is causing errors on input like 1Eb. 

Interestingly it seems to work fine when converting to extremely large numbers (1 google gigabytes converts fine)... so maybe we should support exabytes ?  (It seems that Ruby is being kind and auto converting to numeric values that support larger then long values)

**Comment by ycombinator** [comment]
@jakelandis I added back exabytes support in 8fecb2b. Ready for review again. Thanks!

**Comment by jakelandis** [comment]
@ycombinator  ....found one more issue, looks like the second grouping of digits is being treated as the decimal separator. 

```
filter {
  bytes {
        source => "message"
        target => "result"
        decimal_separator => "."
      }
}
```

```       
"message" => "1,000,000"
 "result" => 1000
```
```
 "message" => "1,000,999"
 "result" => 1001
```

**Comment by ycombinator** [comment]
Thanks @jakelandis. While fixing that bug, I realized that the code is trying to be smart about when to use the `decimal_separator` option. Currently it tries to use that option _only_  if there are exactly 3 digits to the right of the rightmost separator (which could be either `.` or `,`), for disambiguation purposes. Otherwise (if there are more than or less than 3 digits to the right of the rightmost separator), the code just treats the rightmost separator as the decimal separator and everything to its left as one smooshed-together number.

Thinking about this some more, I think this attempt to be smart could actually lead to confusion. I'm thinking it might be better to _always_ use the `decimal_separator` option, defaulting it to `.` (valid choices are `.` and `,`). WDYT?

**Comment by jakelandis** [comment]
> always use the decimal_separator

+1 I originally suggested the "smartness"...but I agree it is probably best not to try to be too smart here.  

Also, can I suggest that based on the decimal_separator, you don't include that value in the list of possible digit group separators, since you _should_ never the same value for digit and decimal separator (i _think_)

EDIT: clarified statement

**Comment by ycombinator** [comment]
> Also, can I suggest that based on the decimal_separator, you don't include that value in the list of possible digit group separators, since you should never the same value for digit and decimal separator (i think)

So another way to think of this would be: don't allow the decimal separator to show up more than once.  Is that right? Also, what should we (Logstash) do in this case?

**Comment by jakelandis** [comment]
> don't allow the decimal separator to show up more than once. Is that right?

yeah, that is logically equivalent

> Also, what should we (Logstash) do in this case?

tag it as a failure and optionally log why at debug.

**Comment by ycombinator** [comment]
@jakelandis This is ready for review again. Thanks!

**Comment by jakelandis** [comment]
Is this the right behavior ?
```
      decimal_separator => ","
```

```
  "message" => "1,000,000"
  "result" => 1,
```
  
I think we should error in this case (more then 1 decimal separator) in case there are is a mix of separators in the data.

**Comment by ycombinator** [comment]
Thanks @jakelandis. I added a [check](https://github.com/logstash-plugins/logstash-filter-bytes/pull/1/commits/3f57b2cc50208fc3324a2ee87163556d01812f47#diff-a11cdef7d1b7142e0a2fdf5bb62f9bddR90) for multiple decimal separators but its happening a bit too late in the execution. I'll fix it shortly.

**Comment by ycombinator** [comment]
@jakelandis Fixed the location of the multiple decimal separators check in the code in f72e78e. Also added a test for your example that caught this issue.

**Comment by jakelandis** [comment]
perfect-1300767_640

LGTM :shipit:

File: lib/logstash/filters/bytes.rb [hunk 26]
```diff
@@ -0,0 +1,116 @@
...
+  PREFIX_POWERS = {
+    'k' => 1, # 1 kilobyte = 1024 ^ 1 bytes
+    'm' => 2, # 1 megabyte = 1024 ^ 2 bytes
+    'g' => 3, # 1 gigabyte = 1024 ^ 3 bytes
+    't' => 4, # 1 terabyte = 1024 ^ 4 bytes
+    'p' => 5, # 1 petabyte = 1024 ^ 5 bytes
+    'e' => 6, # 1 exabyte = 1024 ^ 6 bytes
+  }.freeze
+
+  DIGIT_GROUP_SEPARATORS = " _,."
```
Comment (robbavey) [inline lib/logstash/filters/bytes.rb:20]: Is there potentially a case for making this configurable too (not necessarily in this initial commit)? Digits are commonly separated by spaces in Europe, and by an apostrophe in Switzerland.

File: lib/logstash/filters/bytes.rb [hunk 27]
```diff
@@ -0,0 +1,116 @@
...
+  # Decimal separator
+  config :decimal_separator, :validate => [ ".", "," ] , :default => "."
+
+  # Append values to the `tags` field when there has been no
+  # successful match
+  config :tag_on_failure, :validate => :array, :default => ["_bytesparsefailure"]
+
+  private
+  def normalize_number(number)
+    return number
```
Comment (robbavey) [inline lib/logstash/filters/bytes.rb]: Nit: Unnecessary `return` statement

File: lib/logstash/filters/bytes.rb (hunk 26 above)
Comment (ycombinator) [reply lib/logstash/filters/bytes.rb:20]: ++ to making it configurable. I'd also like to defer that to a follow up PR as this one's pretty large already.

File: spec/filters/bytes_spec.rb [hunk 28]
```diff
@@ -0,0 +1,271 @@
...
+      expect(subject.get('dest')).to eq((3.56 * 1024 * 1024).round)
+    end      
+
+    sample("3,124.56 mb") do
+      expect(subject).to include("dest")
+      expect(subject.get('dest')).to eq((3124.56 * 1024 * 1024).round)
+    end      
+  end
+
+  describe "non-default decimal separator" do
```
Comment (robbavey) [inline spec/filters/bytes_spec.rb:227]: Maybe could do with a couple more tests, particularly using a digit separator as well as a decimal separator.

**Review by robbavey — COMMENTED** [review]
@ycombinator I've left a couple of small nits/comments, but nothing that should block release. Overall, this is a really nice plugin and will be a great addition.

File: spec/filters/bytes_spec.rb (hunk 28 above)
Comment (ycombinator) [reply spec/filters/bytes_spec.rb:227]: I just added a test using a digit separator with a non-default decimal separator in e3bff38. There's already [a test](https://github.com/logstash-plugins/logstash-filter-bytes/pull/1/files#diff-272b5895421d51cf158a043580c1197aR221) using a digit separator with the default decimal separator.
